nats/src/main.rs

```diff
@@ -0,0 +1,302 @@
+            config.cluster_uris.push(url.clone());
+        }
+        if let Some(jwt) = values.get(ENV_NATS_CLIENT_JWT) {
+            config.auth_jwt = Some(jwt.clone());
+        }
+        if let Some(seed) = values.get(ENV_NATS_CLIENT_SEED) {
+            config.auth_seed = Some(seed.clone());
+        }
+        if config.auth_jwt.is_some() && config.auth_seed.is_none() {
+            return Err(RpcError::InvalidParameter(
+                "if you specify jwt, you must also specify a seed".to_string(),
+            ));
+        }
+        if config.cluster_uris.is_empty() {
+            config.cluster_uris.push(DEFAULT_NATS_URI.to_string());
+        }
+        Ok(config)
+    }
+}
+
+/// Nats implementation for wasmcloud:messaging
+#[derive(Default, Clone)]
+struct NatsMessagingProvider {
+    // store nats connection client per actor
+    actors: Arc<RwLock<HashMap<String, Connection>>>,
+}
+/// use default implementations of provider message handlers
+impl ProviderDispatch for NatsMessagingProvider {}
+impl MessagingReceiver for NatsMessagingProvider {}
+
+impl NatsMessagingProvider {
+    /// attempt to connect to nats url (with jwt credentials, if provided)
+    async fn connect(
+        &self,
+        cfg: ConnectionConfig,
+        ld: &LinkDefinition,
+    ) -> Result<Connection, RpcError> {
+        let mut opts = match (cfg.auth_jwt, cfg.auth_seed) {
+            (Some(jwt), Some(seed)) => {
+                let kp = KeyPair::from_seed(&seed)
+                    .map_err(|e| RpcError::ProviderInit(format!("key init: {}", e)))?;
+                asynk::Options::with_jwt(
+                    move || Ok(jwt.clone()),
+                    move |nonce| kp.sign(nonce).unwrap(),
+                )
+            }
+            (None, None) => asynk::Options::new(),
+            _ => {
+                return Err(RpcError::InvalidParameter(
+                    "must provide both jwt and seed for jwt authentication".into(),
+                ));
+            }
+        };
+        opts = opts.with_name("wasmCloud nats-messaging provider");
+        let url = cfg.cluster_uris.get(0).unwrap();
+        let conn = opts
+            .connect(url)
+            .await
+            .map_err(|e| RpcError::ProviderInit(format!("Nats connection to {}: {}", url, e)))?;
+
+        for sub in cfg.subscriptions.iter().filter(|s| !s.is_empty()) {
+            let (sub, queue) = match sub.split_once('|') {
+                Some((sub, queue)) => (sub, Some(queue)),
+                None => (sub.as_str(), None),
+            };
+            self.subscribe(&conn, ld, sub, queue).await?;
+        }
+        Ok(conn)
+    }
+
+    /// send message to subscriber
+    async fn dispatch_msg(&self, ld: &LinkDefinition, nats_msg: NatsMessage) {
+        let msg = SubMessage {
+            body: nats_msg.data,
+            reply_to: nats_msg.reply,
+            subject: nats_msg.subject,
+        };
+        let tx = wasmbus_rpc::provider::ProviderTransport {
```

Comment:
We don't need to do this for MVP, but similar to how the interface was simplified, is there any way to condense the use of the provider transport with the `MessageSubscriberSender`, even if the only thing we do is avoid the manual construction of the struct with a `new`